### PR TITLE
feat: A-star search

### PR DESCRIPTION
**Pull request by kim-em**
This is intended for meta programming, not proving.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
Eventually this is leading to a reimplementation of the original `rewrite_search` from mathlib3, but there's no hurry.

File: Mathlib/Data/ListM/AStar.lean [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    remaining := Γ.nbhd v }
+let σ : State m P V E :=
+  { queue := RBMap.single 0 [v],
+    data := HashMap.empty.insert v d }
+ListM.iterate (do update Γ optimal; readMin)
+  |>.runState' σ |>.map fun d => ⟨d.totalPrio, d.bestPath.to, d.bestPath.edges⟩
+
+/--
+Perform A^* search on a graph,
+starting at a vertex `src` and ending when reaching a vertex satisfying `goal`.
```
Comment (jcommelin) [inline Mathlib/Data/ListM/AStar.lean:224]: Please also describe `optimal` in this docstring.

Should `aStarSearch` and `aStarSearchPaths` refer to each other in their respective docstrings?

Comment (semorrison) [reply Mathlib/Data/ListM/AStar.lean:224]: Yes, good idea!

Done.

File: Mathlib/Data/MLList/AStar.lean [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+variable [Inhabited P] [Inhabited V] [Inhabited E]
+
+/-- This inhabited instance doesn't really make sense,
+but is needed to write `partial` functions below. -/
+instance : Inhabited (VertexData m P V E) :=
+  ⟨{ bestPath := { prio := default, to := default, edges := [] },
+     heuristicPrio := default
+     totalPrio := default
+     explored := []
+     remaining := .nil }⟩
```
Comment (fpvandoorn) [inline Mathlib/Data/MLList/AStar.lean:95]: Maybe just write `deriving Inhabited` after `Path` and `VertexData` instead of doing this manually?

**Review by fpvandoorn — COMMENTED** [review]
one small thing: you use `lowest priority` and `highest priority` synonymously (I think).

bors d+

**Comment by semorrison** [comment]
This was support for a previously planned implementation of `rw_search` that I didn't end up pursuing. I think this code is sufficiently specialised that it would not be useful for others, so I'm going to close.

**Pull request closed without merging**